src/Passes/Source/Validator/Validator.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "ValidationPass/ValidationPass.hpp"
+#include "Validator/Validator.hpp"
+
+#include "Llvm/Llvm.hpp"
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    Validator::Validator(ValidationPassConfiguration const& cfg, bool debug, llvm::TargetMachine* target_machine)
+      : loop_analysis_manager_{debug}
+      , function_analysis_manager_{debug}
+      , gscc_analysis_manager_{debug}
+      , module_analysis_manager_{debug}
+    {
+
+        pass_builder_ = std::make_unique<llvm::PassBuilder>(target_machine);
+
+        pass_builder_->registerModuleAnalyses(module_analysis_manager_);
+        pass_builder_->registerCGSCCAnalyses(gscc_analysis_manager_);
+        pass_builder_->registerFunctionAnalyses(function_analysis_manager_);
+        pass_builder_->registerLoopAnalyses(loop_analysis_manager_);
+
+        pass_builder_->crossRegisterProxies(
+            loop_analysis_manager_, function_analysis_manager_, gscc_analysis_manager_, module_analysis_manager_);
+
+        module_pass_manager_.addPass(ValidationPass(cfg));
+    }
+
+    bool Validator::validate(llvm::Module& module)
+    {
+        llvm::VerifierAnalysis verifier;
+        auto                   result = verifier.run(module, module_analysis_manager_);
+
+        if (result.IRBroken)
+        {
+            llvm::errs() << "; Fatal error: Invalid IR.\n";
+            return false;
+        }
+
+        try
+        {
+            module_pass_manager_.run(module, module_analysis_manager_);
+        }
+        catch (std::exception const& e)
+        {
+            llvm::errs() << "; Fatal error: " << e.what() << "\n";
+            return false;
+        }
+        return true;
+    }
+
+    void Validator::setModulePassManager(llvm::ModulePassManager&& manager)
```

Comment:
Looks like ValidationPassConfiguration passed to the constructor would be gone if setModulePassManager is called. Not sure what it means, but a comment would help.

Reply:
That is a good point. I do not believe that `setModulePassManager` is used at the moment, but either it should be removed or the structure should be reviewed. Unfortunately, I will not be able to do this tonight.